File: answers/leetcode/Number of Islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """

        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if '1' == grid[i][j]:
                    self.dfs(i, j, grid)
                    count += 1
        return count

    def dfs(self, x, y, grid):
        if x < 0 or x >= len(grid) or y < 0 or y >= len(grid[x]) or '1' != grid[x][y]:
            return
        grid[x][y] = '2'
        self.dfs(x - 1, y, grid)
        self.dfs(x + 1, y, grid)
        self.dfs(x, y - 1, grid)
        self.dfs(x, y + 1, grid)

File: answers/other/answers/leetcode/Number of Islands.cpp
//@type  Depth-first Search Breadth-first Search Union Find
//@result 45 / 45 test cases passed. Status: Accepted Runtime: 8 ms Submitted: 0 minutes ago

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int numIslands(vector<vector<char> >& grid) {
		int count(0);
        for (size_t i = 0; i < grid.size(); ++ i) {
			for (size_t j = 0; j < grid[i].size(); ++ j) {
				//cout << "test " << i << " " << j << endl;
				if ('1' == grid[i][j]) {
					++ count;
					dfs(i, j, grid);
				}
			}
		}
		return count;
    }
	void dfs(size_t x, size_t y, vector<vector<char> > &grid) {
		if ('1' == grid[x][y]) {
			grid[x][y] = '0';
			if (x >= 1) {
				dfs(x - 1, y, grid);
			}
			if (x + 1 < grid.size()) {
				dfs(x + 1, y, grid);
			}
			if (y >= 1) {
				dfs(x, y - 1, grid);
			}
			if (y + 1 < grid[x].size()) {
				dfs(x, y + 1, grid);
			}
		}
	}
};

int main() {
	vector<vector<char> > grid;
	vector<char> temp;
	temp.push_back('1');
	temp.push_back('0');
	grid.push_back(temp);
	Solution s;
	cout << s.numIslands(grid) << endl;
	return 0;
}